File: zig/src/examples.zig
const std = @import("std");
const framework = @import("framework.zig");
const agnostic = @import("agnostic.zig");

pub const SearchInput = agnostic.SearchRecord;

pub fn insertionSort(allocator: std.mem.Allocator, input: []const i32) ![]i32 {
    const values = try allocator.alloc(i32, input.len);
    @memcpy(values, input);
    return insertionSortInPlace(allocator, values);
}

pub fn insertionSortInPlace(_: std.mem.Allocator, values: []i32) ![]i32 {
    var index: usize = 1;
    while (index < values.len) : (index += 1) {
        const current = values[index];
        var cursor = index;
        while (cursor > 0 and values[cursor - 1] > current) : (cursor -= 1) {
            values[cursor] = values[cursor - 1];
        }
        values[cursor] = current;
    }

    return values;
}

pub fn mergeSort(allocator: std.mem.Allocator, input: []const i32) ![]i32 {
    if (input.len <= 1) {
        const values = try allocator.alloc(i32, input.len);
        @memcpy(values, input);
        return values;
    }

    const midpoint = input.len / 2;
    const left = try mergeSort(allocator, input[0..midpoint]);
    const right = try mergeSort(allocator, input[midpoint..]);
    return merge(allocator, left, right);
}

pub fn zigSort(allocator: std.mem.Allocator, input: []const i32) ![]i32 {
    const values = try allocator.alloc(i32, input.len);
    @memcpy(values, input);
    std.sort.heap(i32, values, {}, comptime std.sort.asc(i32));
    return values;
}

pub fn binarySearch(_: std.mem.Allocator, input: SearchInput) !i32 {
    var low: usize = 0;
    var high: usize = input.values.len;
    while (low < high) {
        const mid = low + (high - low) / 2;
        const value = input.values[mid];
        if (value == input.target) {
            return @as(i32, @intCast(mid));
        }
        if (value < input.target) {
            low = mid + 1;
        } else {
            high = mid;
        }
    }
    return -1;
}

pub fn linearSearch(_: std.mem.Allocator, input: SearchInput) !i32 {
    for (input.values, 0..) |value, index| {
        if (value == input.target) {
            return @as(i32, @intCast(index));
        }
    }
    return -1;
}

pub fn ExampleSuite(comptime Input: type, comptime Output: type) type {
    return struct {
        adapted: agnostic.AdaptedCases(Input, Output),
        suite: framework.AlgorithmTestSuite(Input, Output),

        pub fn deinit(self: @This()) void {
            self.adapted.deinit();
        }

        pub fn run(self: @This()) !framework.AlgorithmSuiteResult {
            return self.suite.run();
        }
    };
}

pub fn insertionSortSuite(allocator: std.mem.Allocator) !ExampleSuite([]const i32, []i32) {
    return sortingSuite(allocator, "insertion-sort", insertionSort);
}

pub fn mergeSortSuite(allocator: std.mem.Allocator) !ExampleSuite([]const i32, []i32) {
    return sortingSuite(allocator, "merge-sort", mergeSort);
}

pub fn insertionSortInPlaceSuite(allocator: std.mem.Allocator) !ExampleSuite([]i32, []i32) {
    const data = try agnostic.loadSortingCases(allocator);
    errdefer allocator.free(data);
    const cases = try agnostic.adaptSortingInPlace(allocator, data);
    errdefer allocator.free(cases);
    const adapted = agnostic.AdaptedCases([]i32, []i32){
        .data = data,
        .cases = cases,
        .allocator = allocator,
    };
    return .{
        .adapted = adapted,
        .suite = .{
            .suite_name = "sorting",
            .algorithm_name = "insertion-sort-in-place",
            .algorithm = insertionSortInPlace,
            .cases = adapted.cases,
            .options = framework.BenchmarkOptions.quick(),
            .allocator = allocator,
        },
    };
}

pub fn binarySearchSuite(allocator: std.mem.Allocator) !ExampleSuite(SearchInput, i32) {
    const data = try agnostic.loadSearchCases(allocator);
    errdefer allocator.free(data);
    const cases = try agnostic.adaptSearch(allocator, data);
    errdefer allocator.free(cases);
    const adapted = agnostic.AdaptedCases(SearchInput, i32){
        .data = data,
        .cases = cases,
        .allocator = allocator,
    };
    return .{
        .adapted = adapted,
        .suite = .{
            .suite_name = "search",
            .algorithm_name = "binary-search",
            .algorithm = binarySearch,
            .cases = adapted.cases,
            .options = framework.BenchmarkOptions.quick(),
            .allocator = allocator,
        },
    };
}

pub fn runExamples(allocator: std.mem.Allocator) !bool {
    var insertion_suite = try insertionSortSuite(allocator);
    defer insertion_suite.deinit();
    var insertion_result = try insertion_suite.run();
    defer insertion_result.deinit();
    var insertion_in_place_suite = try insertionSortInPlaceSuite(allocator);
    defer insertion_in_place_suite.deinit();
    var insertion_in_place_result = try insertion_in_place_suite.run();
    defer insertion_in_place_result.deinit();
    var merge_suite = try mergeSortSuite(allocator);
    defer merge_suite.deinit();
    var merge_result = try merge_suite.run();
    defer merge_result.deinit();
    var binary_suite = try binarySearchSuite(allocator);
    defer binary_suite.deinit();
    var binary_result = try binary_suite.run();
    defer binary_result.deinit();

    const results = [_]framework.AlgorithmSuiteResult{ insertion_result, insertion_in_place_result, merge_result, binary_result };
    return framework.AlgorithmTestRunner.printReport(&results);
}

fn sortingSuite(
    allocator: std.mem.Allocator,
    algorithm_name: []const u8,
    algorithm: framework.Algorithm([]const i32, []i32),
) !ExampleSuite([]const i32, []i32) {
    const data = try agnostic.loadSortingCases(allocator);
    errdefer allocator.free(data);
    const cases = try agnostic.adaptSortingOutOfPlace(allocator, data);
    errdefer allocator.free(cases);
    const adapted = agnostic.AdaptedCases([]const i32, []i32){
        .data = data,
        .cases = cases,
        .allocator = allocator,
    };
    return .{
        .adapted = adapted,
        .suite = .{
            .suite_name = "sorting",
            .algorithm_name = algorithm_name,
            .algorithm = algorithm,
            .cases = adapted.cases,
            .options = framework.BenchmarkOptions.quick(),
            .allocator = allocator,
        },
    };
}

fn merge(allocator: std.mem.Allocator, left: []const i32, right: []const i32) ![]i32 {
    const merged = try allocator.alloc(i32, left.len + right.len);
    var left_index: usize = 0;
    var right_index: usize = 0;
    var merged_index: usize = 0;

    while (left_index < left.len and right_index < right.len) {
        if (left[left_index] <= right[right_index]) {
            merged[merged_index] = left[left_index];
            left_index += 1;
        } else {
            merged[merged_index] = right[right_index];
            right_index += 1;
        }
        merged_index += 1;
    }
    while (left_index < left.len) : (left_index += 1) {
        merged[merged_index] = left[left_index];
        merged_index += 1;
    }
    while (right_index < right.len) : (right_index += 1) {
        merged[merged_index] = right[right_index];
        merged_index += 1;
    }

    return merged;
}

fn sliceEquality(expected: []i32, actual: []i32) bool {
    return std.mem.eql(i32, expected, actual);
}

fn intEquality(expected: i32, actual: i32) bool {
    return expected == actual;
}

File: python/solid_guide_algorithms/examples.py
from __future__ import annotations

from dataclasses import dataclass
from random import Random

from .framework import (
    Algorithm,
    AlgorithmCase,
    AlgorithmTestRunner,
    AlgorithmTestSuite,
    BenchmarkOptions,
    ComplexityBudget,
)


def insertion_sort(values: list[int]) -> list[int]:
    result = list(values)
    for index in range(1, len(result)):
        current = result[index]
        cursor = index - 1
        while cursor >= 0 and result[cursor] > current:
            result[cursor + 1] = result[cursor]
            cursor -= 1
        result[cursor + 1] = current
    return result


def merge_sort(values: list[int]) -> list[int]:
    if len(values) <= 1:
        return list(values)
    midpoint = len(values) // 2
    return _merge(merge_sort(values[:midpoint]), merge_sort(values[midpoint:]))


def python_sort(values: list[int]) -> list[int]:
    return sorted(values)


@dataclass(frozen=True)
class SearchInput:
    values: tuple[int, ...]
    target: int


def binary_search(input_data: SearchInput) -> int:
    low = 0
    high = len(input_data.values) - 1
    while low <= high:
        mid = low + (high - low) // 2
        value = input_data.values[mid]
        if value == input_data.target:
            return mid
        if value < input_data.target:
            low = mid + 1
        else:
            high = mid - 1
    return -1


def linear_search(input_data: SearchInput) -> int:
    for index, value in enumerate(input_data.values):
        if value == input_data.target:
            return index
    return -1


def insertion_sort_suite() -> AlgorithmTestSuite[list[int], list[int]]:
    return _sorting_suite("insertion-sort", insertion_sort)


def merge_sort_suite() -> AlgorithmTestSuite[list[int], list[int]]:
    return _sorting_suite("merge-sort", merge_sort)


def binary_search_suite() -> AlgorithmTestSuite[SearchInput, int]:
    budget = ComplexityBudget.with_limits(
        max_average_duration_seconds=0.002,
        max_memory_delta_bytes=256 * 1024,
    )
    return AlgorithmTestSuite(
        suite_name="search",
        algorithm_name="binary-search",
        algorithm=binary_search,
        options=BenchmarkOptions.quick(),
        cases=(
            AlgorithmCase(
                name="finds first element",
                input_factory=lambda: SearchInput((1, 3, 5, 7, 9), 1),
                expected_output=linear_search,
                budget=budget,
            ),
            AlgorithmCase(
                name="finds middle element",
                input_factory=lambda: SearchInput((1, 3, 5, 7, 9), 5),
                expected_output=linear_search,
                budget=budget,
            ),
            AlgorithmCase(
                name="reports absent element",
                input_factory=lambda: SearchInput((1, 3, 5, 7, 9), 4),
                expected_output=linear_search,
                budget=budget,
            ),
        ),
    )


def run_examples() -> None:
    AlgorithmTestRunner.run_and_exit(
        insertion_sort_suite().run(),
        merge_sort_suite().run(),
        binary_search_suite().run(),
    )


def _sorting_suite(
    algorithm_name: str,
    algorithm: Algorithm[list[int], list[int]],
) -> AlgorithmTestSuite[list[int], list[int]]:
    small_budget = ComplexityBudget.with_limits(
        max_average_duration_seconds=0.005,
        max_memory_delta_bytes=1024 * 1024,
    )
    return AlgorithmTestSuite(
        suite_name="sorting",
        algorithm_name=algorithm_name,
        algorithm=algorithm,
        options=BenchmarkOptions.quick(),
        cases=(
            AlgorithmCase(
                name="empty input",
                input_factory=list,
                expected_output=python_sort,
                budget=small_budget,
            ),
            AlgorithmCase(
                name="duplicates and negatives",
                input_factory=lambda: [7, -1, 3, 3, 0, -1, 11],
                expected_output=python_sort,
                budget=small_budget,
            ),
            AlgorithmCase(
                name="reverse ordered",
                input_factory=lambda: [9, 8, 7, 6, 5, 4, 3, 2, 1],
                expected_output=python_sort,
                budget=small_budget,
            ),
            AlgorithmCase(
                name="seeded random sample",
                input_factory=lambda: _random_integers(64, 8675309),
                expected_output=python_sort,
                budget=ComplexityBudget.with_limits(
                    max_average_duration_seconds=0.01,
                    max_memory_delta_bytes=2 * 1024 * 1024,
                ),
            ),
        ),
    )


def _merge(left: list[int], right: list[int]) -> list[int]:
    merged: list[int] = []
    left_index = 0
    right_index = 0
    while left_index < len(left) and right_index < len(right):
        if left[left_index] <= right[right_index]:
            merged.append(left[left_index])
            left_index += 1
        else:
            merged.append(right[right_index])
            right_index += 1
    merged.extend(left[left_index:])
    merged.extend(right[right_index:])
    return merged


def _random_integers(count: int, seed: int) -> list[int]:
    random = Random(seed)
    return [random.randrange(-5000, 5000) for _ in range(count)]
